### ISSUE TITLE
Transport PipeOptions

### ISSUE THREAD

**Issue by tmds** [tmds/Tmds.LinuxAsync#25]
ASP.NET Core Socket transport sets `PipeOptions` as follows:
```
            var inputOptions = new PipeOptions(MemoryPool, PipeScheduler.ThreadPool, scheduler, maxReadBufferSize.Value, maxReadBufferSize.Value / 2, useSynchronizationContext: false);
            var outputOptions = new PipeOptions(MemoryPool, scheduler, PipeScheduler.ThreadPool, maxWriteBufferSize.Value, maxWriteBufferSize.Value / 2, useSynchronizationContext: false);
```

The `PipeScheduler.ThreadPool` usage represents whether it is safe to run 'application' code (parsing data, generating more data) on the thread where the socket operation completes.
kestrel-linux-transport (klt) exposes an option: `ApplicationSchedulingMode` and for TE benchmarks this is set to `Inline`. Socket operations complete on the epoll thread of klt (cfr `RunContinuationsAsynchronously=false`), and the application code may run there too.

For the output, there is a non-inline `scheduler` needed. This allows writes coming from the pipes to get coalesced by the time the operation runs on the scheduler.
For Sockets this is the `IOQueue`.
For klt, it is the epoll thread (`DeferSend=true`). Because the epoll thread is used, it is easy to also batch the sends of multiple sockets. There is no way for Sockets transport to use the epoll thread as a scheduler directly, because it is not exposed.

For the input, Sockets uses an `IOQueue` as scheduler.
klt uses `Inline` to configure epoll. Reads are then batched on the epoll thread.
I don't understand why an `IOQueue` is used by Sockets.

@halter73, @davidfowl this is my understanding of PipeOptions and how they differ between Sockets and klt.

cc @adamsitnik

**Comment by tmds** [comment]
> For the input, Sockets uses an IOQueue as scheduler.
klt uses Inline to configure epoll. Reads are then batched on the epoll thread.
I don't understand why an IOQueue is used by Sockets.

@halter73 This scheduler is used when the input pipe was full and the socket-writer is waiting for the app to read it. This doesn't occur in TE benchmarks (requests are small).
When the app has consumed some data, we can do an `Inline` request to the socket to receive more data instead of using the `IOQueue`. Does that make sense?

**Comment by halter73** [comment]
Yeah. Most if not all TE benchmarks don't face PipeWriter.FlushAsync() backpressure in either direction which makes the writer schedulers irrelevant to them.

I don't think this is a reason to make the writer schedulers `Inline` though. Basically that's saying we should call Socket.ReceiveAsync() inline inside PipeReader.AdvanceTo() if we've crossed under the resumeWriterThreshold. This delays the next call to PipeReader.ReadAsync(). Most of the time, I'd expect there's still data left in the Pipe to be read in this case.

I'm not sure it's critical that resuming reads on a Socket where the application has just started catching up consuming the data needs to be done before the other work that was dispatched to the IOQueue previously. And because the pauseWriterThreshold is generally double the resumeWriterThreshold, the writer scheduler should be used far less than the reader scheduler even when there is PipeWriter.FlushAsync() backpressure.

**Comment by tmds** [comment]
These are all the schedulers used:
```cs
_receiver = new SocketReceiver(_socket, A,...
_sender = new SocketSender(_socket, B, ...
var inputOptions = new PipeOptions(MemoryPool, C, D, ...
var outputOptions = new PipeOptions(MemoryPool, E, F, ...

A, B: IOQueue (Windows), Inline (Linux, actually: ThreadPool via SocketEngine)
C, F: ThreadPool
D, E: IOQueue
```

There are two reasons for not using `Inline`:
- We need to use the `ThreadPool`, because user-code may block the thread
- There is a performance advantage by deferring. In case what is deferred is safe (non-blocking), we can use the `IOQueue` because it is faster (due to less contention) than the `ThreadPool`.

For A and B maybe it is advantageous for these to be `Inline`?

For C and F, this is about the untrusted user code. If Kestrel defers input/output at a later point to the ThreadPool, these could be `Inline`
From putting a breakpoint in an http request handler, it seems output is not deferred, so F needs to be `ThreadPool`.
The http handler runs on the ThreadPool even when above schedulers are set to `Inline`, so maybe C could be `Inline`?

E needs to be `IOQueue` for the coalescing of writes.

D is what we discussed, I think there are some reasons to make it `Inline` and some to make it `IOQueue`.

**Comment by tmds** [comment]
> For A and B maybe it is advantageous for these to be Inline?

@adamsitnik can you please run these benchmarks to get an idea of what the effect is?
```
--path "/json" --arg "-e=epoll" --arg "-t=1" --arg "-s=false" --arg "-r=false" --arg "-c=true" --arg "-a=false" --connections 256
--path "/json" --arg "-e=epoll" --arg "-t=1" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --connections 256
--path "/json" --arg "-e=epoll" --arg "-t=1" --arg "-s=false" --arg "-r=false" --arg "-c=true" --arg "-a=true" --connections 256
--path "/json" --arg "-e=epoll" --arg "-t=1" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --connections 256
```
This causes the completions for `SocketAsyncEventArgs` to run directly on the epoll thread.
Additional optimizations may be possible when the completion starts a new operation (like in the receive/send loop).

**Comment by adamsitnik** [comment]
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=1" --arg "-s=false" --arg "-r=false" --arg "-c=true" --arg "-a=false" --connections 256
```
```log
RequestsPerSecond:           723,199
Max CPU (%):                 100
WorkingSet (MB):             406
Avg. Latency (ms):           0.48
Startup (ms):                1
First Request (ms):          52.39
Latency (ms):                0.1
Total Requests:              10,920,324
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             6,502
Published Size (KB):         1,059
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20121.13
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=1" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --connections 256
```
```log
RequestsPerSecond:           212,388
Max CPU (%):                 50
WorkingSet (MB):             383
Avg. Latency (ms):           1.21
Startup (ms):                1
First Request (ms):          53.75
Latency (ms):                0.11
Total Requests:              3,206,938
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,001
Published Size (KB):         1,059
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20121.13
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=1" --arg "-s=false" --arg "-r=false" --arg "-c=true" --arg "-a=true" --connections 256
```
```log
RequestsPerSecond:           813,332
Max CPU (%):                 100
WorkingSet (MB):             404
Avg. Latency (ms):           0.5
Startup (ms):                1
First Request (ms):          53.97
Latency (ms):                0.11
Total Requests:              12,280,684
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,001
Published Size (KB):         1,059
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20121.13
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=1" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --connections 256
```
```log
RequestsPerSecond:           209,409
Max CPU (%):                 50
WorkingSet (MB):             387
Avg. Latency (ms):           1.23
Startup (ms):                1
First Request (ms):          53.87
Latency (ms):                0.09
Total Requests:              3,161,934
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,001
Published Size (KB):         1,059
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20121.13
```

**Comment by tmds** [comment]
@adamsitnik can you try incrementing `t` for this benchmark until we're at 100% cpu?
```
--path "/json" --arg "-e=epoll" --arg "-t=1" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --connections 256
```

**Comment by tmds** [comment]
> For A and B maybe it is advantageous for these to be Inline?

As Adam observed, AIO was not being used in these benchmarks (https://github.com/tmds/Tmds.LinuxAsync/issues/36). This is due to the way SocketConnection splits a receive in `WaitForDataAsync` and `ReceiveAsync` (see https://github.com/tmds/Tmds.LinuxAsync/issues/36#issuecomment-590380728).
To be able to batch in such a configuration, A needs to be `Inline`.

At optimal threadcount, we measured a 10.6% improvement for JSON benchmark (https://github.com/tmds/Tmds.LinuxAsync/issues/36#issuecomment-591348906).

On Windows setting A and B to `Inline` may also have a positive effect on perf.

**Comment by tmds** [comment]
I've added a few options in https://github.com/tmds/Tmds.LinuxAsync/pull/38. Let's measure how they affect performance for TE JSON,
- Complete receives on epoll thread instead of ThreadPool (c=false) 
- (new) Trust app code (i=true), use Inline instead of ThreadPool
- (new) Don't coalesce writes (p=false), use Inline instead of IOQueue
- (new) Don't wait for data available (w=false)

**note**: for `c=false`, `t` is set to `4` to avoid the epoll thread becoming the bottleneck.

@adamsitnik, can you run these benchmarks?

```
// Without AIO
--path "/json" --arg "-e=epoll" --arg "-t=1" --arg "-s=false" --arg "-r=false" --arg "-c=true" --arg "-a=false" --connections 256
--path "/json" --arg "-e=epoll" --arg "-t=4" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --connections 256
--path "/json" --arg "-e=epoll" --arg "-t=4" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --arg "-i=true" --connections 256
--path "/json" --arg "-e=epoll" --arg "-t=4" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --arg "-p=false" --connections 256
--path "/json" --arg "-e=epoll" --arg "-t=4" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --arg "-w=false" --connections 256

// With AIO
--path "/json" --arg "-e=epoll" --arg "-t=4" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --connections 256
--path "/json" --arg "-e=epoll" --arg "-t=4" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --arg "-i=true" --connections 256
--path "/json" --arg "-e=epoll" --arg "-t=4" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --arg "-p=false" --connections 256
--path "/json" --arg "-e=epoll" --arg "-t=4" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --arg "-w=false" --connections 256
```
```

**Comment by adamsitnik** [comment]
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=1" --arg "-s=false" --arg "-r=false" --arg "-c=true" --arg "-a=false" --connections 256
```
```log
RequestsPerSecond:           394,203
Max CPU (%):                 98
WorkingSet (MB):             195
Avg. Latency (ms):           0.93
Startup (ms):                2
First Request (ms):          89.25
Latency (ms):                0.56
Total Requests:              5,951,925
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,002
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=4" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --connections 256
```
```log
RequestsPerSecond:           404,606
Max CPU (%):                 100
WorkingSet (MB):             186
Avg. Latency (ms):           1.86
Startup (ms):                2
First Request (ms):          85.98
Latency (ms):                0.63
Total Requests:              6,108,690
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,002
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=4" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --arg "-i=true" --connections 256
```
```log
RequestsPerSecond:           180,593
Max CPU (%):                 82
WorkingSet (MB):             194
Avg. Latency (ms):           1.47
Startup (ms):                2
First Request (ms):          93.37
Latency (ms):                0.52
Total Requests:              2,726,919
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,002
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=4" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --arg "-p=false" --connections 256
```
```log
RequestsPerSecond:           363,112
Max CPU (%):                 100
WorkingSet (MB):             196
Avg. Latency (ms):           1.61
Startup (ms):                2
First Request (ms):          85.11
Latency (ms):                0.35
Total Requests:              5,482,910
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,002
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=4" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --arg "-w=false" --connections 256
```
```log
RequestsPerSecond:           432,472
Max CPU (%):                 100
WorkingSet (MB):             185
Avg. Latency (ms):           2
Startup (ms):                1
First Request (ms):          93.83
Latency (ms):                0.67
Total Requests:              6,521,476
Duration: (ms)               15,080
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,002
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=4" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --connections 256
```
```log
RequestsPerSecond:           424,429
Max CPU (%):                 100
WorkingSet (MB):             193
Avg. Latency (ms):           1.69
Startup (ms):                2
First Request (ms):          94.68
Latency (ms):                0.62
Total Requests:              6,407,127
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,002
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=4" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --arg "-i=true" --connections 256
```
```log
RequestsPerSecond:           175,013
Max CPU (%):                 81
WorkingSet (MB):             190
Avg. Latency (ms):           1.5
Startup (ms):                2
First Request (ms):          93.59
Latency (ms):                0.59
Total Requests:              2,642,683
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,002
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=4" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --arg "-p=false" --connections 256
```
```log
RequestsPerSecond:           367,681
Max CPU (%):                 100
WorkingSet (MB):             176
Avg. Latency (ms):           1.67
Startup (ms):                2
First Request (ms):          88.56
Latency (ms):                0.46
Total Requests:              5,551,967
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,002
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=4" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --arg "-w=false" --connections 256
```
```log
RequestsPerSecond:           465,125
Max CPU (%):                 99
WorkingSet (MB):             193
Avg. Latency (ms):           0.93
Startup (ms):                1
First Request (ms):          94.7
Latency (ms):                0.57
Total Requests:              7,023,010
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,002
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```

**Comment by tmds** [comment]
Results in a table:

  | AIO off | Max CPU | AIO on | Max CPU
-- | -- | -- | -- | --
complete on threadpool | 394203 | 98 |   |  
complete on epoll thread | 404606 | 100 | 424429 | 100
inline app code | 180593 | 82 | 175013 | 81
don’t coalesce writes | 363112 | 100 | 367681 | 100
don’t wait for data available | 432472 | 100 | 465125 | 99

Observations:
- Also without AIO we see improved performance by completing on the epoll thread (+2.6%).
- Not waiting for data available gives a nice perf boost with and without AOI.
- Not coalescing writes reduces performance. I guess this is caused by multiple pipe writes for a single json reply (cfr https://github.com/dotnet/runtime/issues/25059).
- Inline app code is unable to max CPU. I guess this moved request parsing to the epoll thread which is now a bottleneck.

@adamsitnik let's try to max CPU also for Inline app code:
```
// Without AIO
--path "/json" --arg "-e=epoll" --arg "-t=5" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --arg "-i=true" --connections 256
--path "/json" --arg "-e=epoll" --arg "-t=6" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --arg "-i=true" --connections 256
--path "/json" --arg "-e=epoll" --arg "-t=7" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --arg "-i=true" --connections 256
--path "/json" --arg "-e=epoll" --arg "-t=8" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --arg "-i=true" --connections 256

// With AIO
--path "/json" --arg "-e=epoll" --arg "-t=5" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --arg "-i=true" --connections 256
--path "/json" --arg "-e=epoll" --arg "-t=6" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --arg "-i=true" --connections 256
--path "/json" --arg "-e=epoll" --arg "-t=7" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --arg "-i=true" --connections 256
--path "/json" --arg "-e=epoll" --arg "-t=8" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --arg "-i=true" --connections 256
```

**Comment by adamsitnik** [comment]
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=5" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --arg "-i=true" --connections 256
```
```log
RequestsPerSecond:           196,972
Max CPU (%):                 90
WorkingSet (MB):             187
Avg. Latency (ms):           1.36
Startup (ms):                1
First Request (ms):          92.84
Latency (ms):                0.62
Total Requests:              2,974,237
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             8,002
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=6" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --arg "-i=true" --connections 256
```
```log
RequestsPerSecond:           193,458
Max CPU (%):                 95
WorkingSet (MB):             206
Avg. Latency (ms):           1.52
Startup (ms):                1
First Request (ms):          94.72
Latency (ms):                0.37
Total Requests:              2,921,130
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,502
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=7" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --arg "-i=true" --connections 256
```
```log
RequestsPerSecond:           220,327
Max CPU (%):                 97
WorkingSet (MB):             199
Avg. Latency (ms):           1.32
Startup (ms):                1
First Request (ms):          92.88
Latency (ms):                0.39
Total Requests:              3,326,963
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,502
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=8" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --arg "-i=true" --connections 256
```
```log
RequestsPerSecond:           223,899
Max CPU (%):                 99
WorkingSet (MB):             198
Avg. Latency (ms):           1.29
Startup (ms):                1
First Request (ms):          88.82
Latency (ms):                0.37
Total Requests:              3,380,875
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,502
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=5" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --arg "-i=true" --connections 256
```
```log
RequestsPerSecond:           181,266
Max CPU (%):                 89
WorkingSet (MB):             206
Avg. Latency (ms):           1.49
Startup (ms):                1
First Request (ms):          94.1
Latency (ms):                0.33
Total Requests:              2,737,050
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,002
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=6" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --arg "-i=true" --connections 256
```
```log
RequestsPerSecond:           198,395
Max CPU (%):                 94
WorkingSet (MB):             203
Avg. Latency (ms):           1.4
Startup (ms):                1
First Request (ms):          94.41
Latency (ms):                0.45
Total Requests:              2,995,691
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             5,502
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=7" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --arg "-i=true" --connections 256
```
```log
RequestsPerSecond:           215,301
Max CPU (%):                 97
WorkingSet (MB):             205
Avg. Latency (ms):           1.34
Startup (ms):                1
First Request (ms):          92.98
Latency (ms):                0.71
Total Requests:              3,250,852
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,002
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=8" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --arg "-i=true" --connections 256
```
```log
RequestsPerSecond:           226,645
Max CPU (%):                 98
WorkingSet (MB):             207
Avg. Latency (ms):           1.23
Startup (ms):                1
First Request (ms):          93.39
Latency (ms):                0.63
Total Requests:              3,422,319
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,001
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```

**Comment by tmds** [comment]
Thanks!

Looks like we should go up even higher: adding more threads still improves RPS.
Going up till 12, that is `ProcessorCount` on this machine, right?

```
// Without AIO
--path "/json" --arg "-e=epoll" --arg "-t=9" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --arg "-i=true" --connections 256
--path "/json" --arg "-e=epoll" --arg "-t=10" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --arg "-i=true" --connections 256
--path "/json" --arg "-e=epoll" --arg "-t=11" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --arg "-i=true" --connections 256
--path "/json" --arg "-e=epoll" --arg "-t=12" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --arg "-i=true" --connections 256

// With AIO
--path "/json" --arg "-e=epoll" --arg "-t=9" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --arg "-i=true" --connections 256
--path "/json" --arg "-e=epoll" --arg "-t=10" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --arg "-i=true" --connections 256
--path "/json" --arg "-e=epoll" --arg "-t=11" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --arg "-i=true" --connections 256
--path "/json" --arg "-e=epoll" --arg "-t=12" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --arg "-i=true" --connections 256
```

**Comment by adamsitnik** [comment]
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=9" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --arg "-i=true" --connections 256
```
```log
RequestsPerSecond:           230,187
Max CPU (%):                 99
WorkingSet (MB):             196
Avg. Latency (ms):           1.3
Startup (ms):                1
First Request (ms):          95.65
Latency (ms):                0.44
Total Requests:              3,475,640
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             6,502
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=10" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --arg "-i=true" --connections 256
```
```log
RequestsPerSecond:           239,949
Max CPU (%):                 100
WorkingSet (MB):             182
Avg. Latency (ms):           1.2
Startup (ms):                1
First Request (ms):          93.81
Latency (ms):                0.36
Total Requests:              3,623,152
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,001
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=11" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --arg "-i=true" --connections 256
```
```log
RequestsPerSecond:           252,246
Max CPU (%):                 100
WorkingSet (MB):             200
Avg. Latency (ms):           1.16
Startup (ms):                1
First Request (ms):          94.96
Latency (ms):                0.36
Total Requests:              3,808,858
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,502
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=12" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=false" --arg "-i=true" --connections 256
```
```log
RequestsPerSecond:           260,546
Max CPU (%):                 100
WorkingSet (MB):             193
Avg. Latency (ms):           1.11
Startup (ms):                1
First Request (ms):          92.73
Latency (ms):                0.6
Total Requests:              3,934,155
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,001
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=9" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --arg "-i=true" --connections 256
```
```log
RequestsPerSecond:           233,359
Max CPU (%):                 99
WorkingSet (MB):             197
Avg. Latency (ms):           1.23
Startup (ms):                1
First Request (ms):          92.01
Latency (ms):                0.52
Total Requests:              3,523,536
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,001
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=10" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --arg "-i=true" --connections 256
```
```log
RequestsPerSecond:           252,141
Max CPU (%):                 100
WorkingSet (MB):             195
Avg. Latency (ms):           1.18
Startup (ms):                1
First Request (ms):          92.58
Latency (ms):                0.7
Total Requests:              3,807,329
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,002
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=11" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --arg "-i=true" --connections 256
```
```log
RequestsPerSecond:           253,025
Max CPU (%):                 100
WorkingSet (MB):             179
Avg. Latency (ms):           1.17
Startup (ms):                1
First Request (ms):          94.71
Latency (ms):                0.31
Total Requests:              3,820,584
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,002
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```
```cmd
--path "/json" --arg "-e=epoll" --arg "-t=12" --arg "-s=false" --arg "-r=false" --arg "-c=false" --arg "-a=true" --arg "-i=true" --connections 256
```
```log
RequestsPerSecond:           256,985
Max CPU (%):                 100
WorkingSet (MB):             195
Avg. Latency (ms):           1.14
Startup (ms):                1
First Request (ms):          93.68
Latency (ms):                0.29
Total Requests:              3,880,355
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             5,002
Published Size (KB):         1,060
SDK:                         5.0.100-preview.1.20106.1
Runtime:                     5.0.0-preview.2.20120.8
ASP.NET Core:                5.0.0-preview.2.20125.12
```

**Comment by tmds** [comment]
I'm surprised inlining the application code doesn't get us at the same level of the other benchmarks with `t=ProcessorCount`.

**Comment by tmds** [comment]
I've opened PRs in in aspnetcore repo with changes to consider.

I'm going to create two issues to look into:

- Verify
> Not coalescing writes reduces performance. I guess this is caused by multiple pipe writes for a single json reply (cfr dotnet/runtime#25059).
- Investigate
> I'm surprised inlining the application code doesn't get us at the same level of the other benchmarks with t=ProcessorCount.
